File: 백준/Silver/15650. N과 M （2）/N과 M （2）.js
const fs = require("fs");
const path = process.platform === "linux" ? "/dev/stdin" : "./input.txt";

const input = fs
  .readFileSync(path, "utf8")
  .split("\n")
  .map((item) => item.trim());

const [N, M] = input[0].split(" ").map((item) => parseInt(item));

function backtrack(current, start) {
  if (current.length === M) {
    console.log(current.join(" "));
    return;
  }

  for (let i = start; i <= N; i++) {
    current.push(i);
    backtrack(current, i + 1);
    current.pop();
  }
}

backtrack([], 1);

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split(' '))
visited = [False for _ in range(N + 1)]
result = []

def dfs(k):
    if len(result) == M:
        print(*result)
        return

    for i in range(k, N + 1):
        if not visited[i]:
            visited[i] = True
            result.append(i)

            dfs(i + 1)

            visited[i] = False
            result.pop()

dfs(1)